### PR TITLE
Domains: Deduplicate code for availability messages

### PR DESCRIPTION
**Pull request by klimeryk**
This has created a lot of problems in the past (for example, #10176).
This change deduplicates the code responsible for handling the availability messages. Additionally, it removes a call to `is-mappable` endpoint - we are already calling the superior `is-available` endpoint
(that properly handles .co subdomains, etc.).

_Additionally_, I've fixed an error that prevented registering a domain from the map-domain-step 😅 

### Testing
Make sure that all the cases returned by the `is-availability` endpoint are properly handled in both register-domain-step (domain search) and map-domain-step (domain mapping). Be sure to test from both domain management and signup.

Additionally, check that it is now possible to add a domain registration directly from the domain mapping view (if you search for an available domain there).

### CODE REVIEW

File: client/my-sites/upgrades/map-domain/index.jsx [hunk 1]
```diff
@@ -82,7 +82,7 @@ var MapDomain = React.createClass( {
 		upgradesActions.addItem(
 			cartItems.domainRegistration( {
 				productSlug: suggestion.product_slug,
-				domain: suggestion.domain
+				domain: suggestion.domain_name
```
Comment (klimeryk) [inline client/my-sites/upgrades/map-domain/index.jsx]: This was causing an error when trying to add a domain registration from the domain mapping view - the proper key is `domain_name`, not `domain`.

Comment (umurkontaci) [inline client/my-sites/upgrades/map-domain/index.jsx]: I think signup might be using the key `domain`

**Comment by umurkontaci** [comment]
Is it possible to split the error messages depending on endpoint?

For example in `register-domain-step.jsx`, we are using the same error handler for both availability and search endpoints and merging the errors. It's impossible to differentiate which endpoint caused the error without going back to backend code.

File: client/components/domains/map-domain-step/index.jsx [hunk 2]
```diff
@@ -164,105 +164,21 @@ const MapDomainStep = React.createClass( {
...
-			}
-
-			return this.props.onMapDomain( domain );
-		} );
-
 		canRegister( domain, ( error, result ) => {
 			if ( error ) {
+				if ( error.code === 'not_available_but_mappable' ) {
+					this.props.onMapDomain( domain );
+					return;
```
Comment (umurkontaci) [inline client/components/domains/map-domain-step/index.jsx]: Why does a function called `canRegister` calls `props.onMapDomain` if the domain is already registered?

File: client/my-sites/upgrades/map-domain/index.jsx (hunk 1 above)
Comment (klimeryk) [inline client/my-sites/upgrades/map-domain/index.jsx]: No, it's using the same format. But, please, feel free to test yourself :)

File: client/components/domains/map-domain-step/index.jsx (hunk 2 above)
Comment (klimeryk) [inline client/components/domains/map-domain-step/index.jsx]: Not sure if this is a rhetorical question or you did not read the code? `canRegister` is a poorly named call to the `is-available` endpoint, which confusingly returns three flags: `is_available`, `is_mappable` and `is_registrable`. Now, to make matters worse, these flags are converted into string errors and passed here in `error`... So, yeah, the code is more confusing than it should but we should untangle that one bit at a time.
But to make this part clearer, I've renamed the `canRegister` import to `checkDomainAvailability` - hope that make it clearer.
As to why it calls `onMapDomain` when a domain is registered - that's the prerequisite for mapping a domain. You can't map a domain that has not been registered.

**Comment by klimeryk** [comment]
> Is it possible to split the error messages depending on endpoint?
> 
> For example in register-domain-step.jsx, we are using the same error handler for both availability and search endpoints and merging the errors. It's impossible to differentiate which endpoint caused the error without going back to backend code.

Sorry, don't understand this. The errors are coming from the same endpoint: `is-available`.

**Comment by umurkontaci** [comment]
No, in `register-domain-step`, we are calling this function from three points:

https://github.com/Automattic/wp-calypso/blob/3db9d689e688dacdfc01f777710d738e9d8f78ae/client/components/domains/register-domain-step/index.jsx#L361

https://github.com/Automattic/wp-calypso/blob/3db9d689e688dacdfc01f777710d738e9d8f78ae/client/components/domains/register-domain-step/index.jsx#L319

https://github.com/Automattic/wp-calypso/blob/3db9d689e688dacdfc01f777710d738e9d8f78ae/client/components/domains/register-domain-step/index.jsx#L170


This function has become an god function of error messages, as the function handles error messages from both of the endpoints (is-available and suggestions), it's impossible to know which is which.

**Comment by klimeryk** [comment]
Ah, you're right - my bad, I hadn't noticed that. But it's been a god function of error messages before my change. I did not add any new ones to it - I just removed the duplicated ones in `map-domain-step`. Which is the point of this PR. I agree that it's still a mess, but I'd prefer to handle it one step at a time. The PR is already large-ish, and I'd prefer not to add unrelated changes to it - it was meant to remove the duplicated messages and share the logic for handling these errors between `register-domain-step` and `map-domain-step` - and it's done that, I think.

File: client/components/domains/map-domain-step/index.jsx [hunk 3]
```diff
@@ -9,12 +9,12 @@ import { localize } from 'i18n-calypso';
  * Internal dependencies
  */
 import { cartItems } from 'lib/cart-values';
-import { getFixedDomainSearch, canMap, canRegister, getTld } from 'lib/domains';
+import { getFixedDomainSearch, canRegister as checkDomainAvailability } from 'lib/domains';
```
Comment (umurkontaci) [inline client/components/domains/map-domain-step/index.jsx]: We should change the original function's name instead of creating aliases

**Comment by umurkontaci** [comment]
Functionality ✅ 

We should clear out the name ambiguity (`canRegister`) though.

**Comment by klimeryk** [comment]
Thanks for the feedback - I've renamed `canRegister` globally to `checkDomainAvailability`.

**Pull request merged**